reactioncommerce/reaction
# #4384 Improve app startup time when large number of Accounts/Users exists

Resolves #4384  
Impact: **major**  
Type: **performance**

## Issue
When an RC install with a large number of Accounts/Users starts up, currently every user is loaded in a single query and updated in another query, to add roles to their accounts. The core function in question - `addRolesToGroups` - is called 3 times throughout the app and is intended to add new roles to groups and their associated accounts/users. Even if the roles passed to the function call are already in the groups passed, all Accounts are loaded and updated.

## Solution
The solution: 1) make it so Groups and Accounts/Users are only updated if the given roles aren't already present in the group, 2) batch the Accounts find query and Users update query to update a maximum of 100k Accounts/Users per call (to avoid a RangeError), and 3) have the heaviest queries run asynchronously in the background so that the app can finish starting up and start serving user requests.

There is one catch: With larger datasets (2,100,000 Accounts/Users tested), although the heaviest queries are run in the background, and the app is able to load, it is noticeably slower than usual while clicking through pages, due to the Mongo load. On the other hand, this large processing is now only done on the first startup. Subsequent startups no longer add roles to 2.1m accounts, because they already exist.

## Breaking changes
None


## Testing
1. Use reaction-data-loader to load the `ret` (retailer) dataset
2. Start the app with `export METEOR_DISABLE_OPTIMISTIC_CACHING=1 && REACTION_LOG_LEVEL="DEBUG" reaction`
3. Confirm in logs that "Reaction initialization finished" takes less less than 15s
4. Confirm you can visit app in browser and click around
5. Confirm you see debug messages about the group batch processing. Ex: `Processing user role update #2 of 21 for customer group`, but can still use the app
6. When finished processing, restart the app
7. Confirm you see messages such as `Skipping roles already assigned to...`, and that once the app starts up, there are no `Processing user role updates...` messages
8. Confirm roles are still added to correct groups and users
